accepted/2020/linking-libraries.md

```diff
@@ -1,170 +1,572 @@
+
+For example: [Annotating startup hook functionality as incompatible with trimming](https://github.com/dotnet/runtime/pull/44050/files#diff-9c65180e471f0d5b540d61b5aee02e5f6992374bef257ef165ac44099f75dfd4L98-L99)
+
+#### Separate incompatible functionality
+
+The next step in improving the experience is to enable at least some subset
+of the functionality while trimming. Frequently only some parts of a feature
+are incompatible with trimming while other parts can still be used without issues.
+To resolve this the feature and often also its public API needs to be changed
+such that static analysis can determine if the incompatible functionality
+is in use or not.
+
+If this is not possible the incompatible functionality can also be disabled
+using [feature switches](#Feature-Switches). See a separate section below about
+more details on this approach.
+
+Example using the feature switch approach: [Introducing the `EventSourceSupported`
+feature switch](https://github.com/dotnet/runtime/commit/a547d4178cd2d71d9b6a7a99600e20a3211d4436)
+
+#### Making the feature truly compatible with trimming
+
+Finally, invest into built-time tooling (or other approaches) to enable usage of
+the full feature even for trimmed apps. For example,
+[source generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/)
+can be used to help with this effort.
+
+It’s expected that some features are never fully trimming compatible.
+Such features should be at the very least annotated with `RequiresUnreferencedCode`
+to provide a good user experience, explaining what feature is incompatible
+and offering potential alternatives.
+
+Example of applying source generators: [Plan to use source generator for `System.Text.Json`](https://github.com/dotnet/designs/blob/main/accepted/2020/serializer/SerializerGoals5.0.md#comparison-of-options)
+
+#### .NET Libraries challenges
+
+There are features in the .NET Libraries that will be very expensive, and in some
+cases virtually impossible, to make 100% trim compatible. A few examples are:
+
+* `System.ComponentModel.Composition`, `System.Composition`, and other dependency
+injection style libraries
+    * Contains features such as loading "parts" from a directory
+* `Microsoft.CSharp` and `System.Linq.Expressions`
+    * Use Reflection extensively to implement `dynamic` and expression trees
+* `Assembly.Load/LoadFrom/LoadFile`
+    - Dynamically loading assemblies into a trimmed application is problematic
+because if the trim tool isn't aware of those assemblies, it won't know to preserve
+the methods required by those assemblies. If the core application doesn't need
+`string.Split`, trimming will remove the `Split` method. When an assembly that
+the trim tool didn't know about is loaded dynamically, it may need `string.Split`,
+and fail to run.
+* `System.Reflection.Emit`
+    * A method or property that is only called by dynamically generated code may
+be trimmed if the linker isn't aware of it.
```

Comment:
This seems like too broad a category to include here.  There are plenty of S.R.Emit solutions that can be appropriately annotated to teach the linker what APIs are actually used, and in many cases, follow patterns with hardcoded reflection the linker already understands.

Reply:
I removed the RefEmit item as it's not a good example. As you mention in lot of cases it is actually perfectly compatible with trimming, or can be made that way without to much trouble.